helix-term/src/ui/picker.rs

```diff
@@ -124,375 +117,147 @@ impl Preview<'_, '_> {
     }
 }
 
-impl<T: Item + 'static> FilePicker<T> {
-    pub fn new(
-        options: Vec<T>,
-        editor_data: T::Data,
-        callback_fn: impl Fn(&mut Context, &T, Action) + 'static,
-        preview_fn: impl Fn(&Editor, &T) -> Option<FileLocation> + 'static,
-    ) -> Self {
-        let truncate_start = true;
-        let mut picker = Picker::new(options, editor_data, callback_fn);
-        picker.truncate_start = truncate_start;
+type PickerCallback<T> = Box<dyn Fn(&mut Context, &T, Action)>;
+
+pub struct Column<T> {
+    // This will be used when we start rendering the table header for pickers with >1 column.
+    #[allow(dead_code)]
+    name: &'static str,
+    // TODO: currently these are dead code. Ideally DynPickers could use a
+    // "display only" column for the entry of their query so that we wouldn't
+    // exclude valid results:
+    // * https://github.com/helix-editor/helix/issues/5714
+    // * https://github.com/helix-editor/helix/pull/4687#issuecomment-1364586418
+    // Then Picker would look for the first column with 'filter: true' and use
+    // that for filtering and same for sorting.
+    sort: bool,
+    filter: bool,
+    format_fn: Box<dyn Fn(&T) -> Cell>,
+    sort_text_fn: Option<Box<dyn Fn(&T) -> Cow<str>>>,
+    filter_text_fn: Option<Box<dyn Fn(&T) -> Cow<str>>>,
+}
```

Comment:
Column is currently implemented as a struct since it's a nice API to write out your set of columns when creating a Picker: https://github.com/helix-editor/helix/blob/b06482654f59c989ffff53cb8ebebd865d31679c/helix-term/src/commands/lsp.rs#L255-L279

`Item::Data` can be eliminated since it can be captured in these closures.

I think it should be possible to re-implement Column as a trait similar to the current `ui::menu::Item`: https://github.com/helix-editor/helix/blob/78e869542022638af482c0dd85a07086a942907a/helix-term/src/ui/menu.rs#L17-L32

```rust
pub trait Column<T> {
    type Data;
    const NAME: &'static str;
    const FILTER: bool;
    const SORT: bool;

    fn format<'a>(&self, item: &'a T, data: &Self::Data) -> Cell<'a>;

    fn sort_text<'a>(&self, item: &'a T, data: &Self::Data) -> Cow<'a, str> {
        let text: String = self.format(data).contents.into();
        text.into()
    }

    fn filter_text<'a>(&self, item: &'a T, data: &Self::Data) -> Cow<'a, str> {
        let text: String = self.format(data).contents.into();
        text.into()
    }
}
```

But it will be more verbose to define columns and create Pickers (`Picker::new` will still need to take `editor_data`).

I don't have a good intuition about the penalty of dynamic dispatch. Is it worth the trait for static dispatch? These functions may be called very often (per-keypress, per-option).

Reply:
I'll switch to the trait version. Not only do we pay the dynamic dispatch penalty but we need to allocate in more situations because of a limitation of closures: you can't return references of captured variables. So we end up allocating `Cow::Owned` in some places where we previously got away with `Cow::Borrowed`s. The trait version will also allow us to re-use some column definitions like we currently do with the `menu::Item` implementation for `PathBuf`.